File: 2017/01/python_day01/day01.py
#!/usr/bin/env python3
"""
Advent of Code 2017 Day 01.
"""


class Day01:
    """Main module for solving Day01."""

    @staticmethod
    def rotate(input_str: str, num: int) -> str:
        """Rotate a string by `num` characters."""
        return input_str[num:] + input_str[:num]

    @staticmethod
    def part1(input_str: str) -> int:
        return Day01.captcha(input_str, 1)

    @staticmethod
    def part2(input_str: str) -> int:
        rotate_amount = int(len(input_str) / 2)
        return Day01.captcha(input_str, rotate_amount)

    @staticmethod
    def captcha(input_str: str, rotate_amount: int) -> int:
        rotated = Day01.rotate(input_str, rotate_amount)
        count = 0
        for str1, str2 in zip(input_str, rotated):
            if str1 == str2:
                count += int(str1)
        return count


if __name__ == "__main__":
    print("Hello from main")
    with open("../input.txt", "r") as file:
        data = file.read().replace("\n", "")
    print("part 1:")
    print(Day01.part1(data))
    print("part 2:")
    print(Day01.part2(data))
    # print("rotate:")
    # r = Day01.rotate("1234", 5)
    # print(r)

File: 2017/01/ruby_day01/lib/day01.rb
class Day01
  def self.part1(nums)
    rotate_amount = 1
    captcha(nums, rotate_amount)
  end

  def self.part2(nums)
    rotate_amount = nums.size / 2
    captcha(nums, rotate_amount)
  end

  def self.captcha(nums, rotate_amount)
    s1 = nums.chars
    s2 = nums.chars.rotate(rotate_amount)
    count = 0
    s1.zip(s2).each { |pair| count += pair[0].to_i if pair[0] == pair[1] }
    count
  end
end
